plugins/graph_algorithm/src/algorithms/neighborhood.cpp

```diff
@@ -0,0 +1,142 @@
+        {
+            if (!graph)
+            {
+                return ERR("graph is a nullptr");
+            }
+
+            if (start_vertices.empty())
+            {
+                return ERR("no start vertices provided");
+            }
+
+            igraph_vector_int_t i_gates;
+            if (auto res = igraph_vector_int_init(&i_gates, start_vertices.size()); res != IGRAPH_SUCCESS)
+            {
+                return ERR(igraph_strerror(res));
+            }
+
+            for (u32 i = 0; i < start_vertices.size(); i++)
+            {
+                VECTOR(i_gates)[i] = start_vertices.at(i);
+            }
+
+            auto res = get_neighborhood_igraph(graph, &i_gates, order, direction, min_dist);
+
+            igraph_vector_int_destroy(&i_gates);
+
+            if (res.is_error())
+            {
+                return ERR(res.get_error());
+            }
+
+            return res;
+        }
+
+        Result<std::vector<std::vector<u32>>> get_neighborhood_igraph(NetlistGraph* graph, const igraph_vector_int_t* start_gates, u32 order, NetlistGraph::Direction direction, u32 min_dist)
+        {
+            if (!graph)
+            {
+                return ERR("graph is a nullptr");
+            }
+
+            igraph_vs_t v_sel;
+            if (auto res = igraph_vs_vector(&v_sel, start_gates); res != IGRAPH_SUCCESS)
+            {
+                return ERR(igraph_strerror(res));
+            }
+
+            igraph_neimode_t mode;
+            switch (direction)
+            {
+                case NetlistGraph::Direction::IN:
+                    mode = IGRAPH_IN;
+                    break;
+                case NetlistGraph::Direction::OUT:
+                    mode = IGRAPH_OUT;
+                    break;
+                case NetlistGraph::Direction::ALL:
+                    mode = IGRAPH_ALL;
+                    break;
+                case NetlistGraph::Direction::NONE:
+                    igraph_vs_destroy(&v_sel);
+                    return ERR("invalid direction 'NONE'");
+            }
+
+            igraph_vector_int_list_t neighborhoods_raw;
+            if (auto res = igraph_vector_int_list_init(&neighborhoods_raw, 1); res != IGRAPH_SUCCESS)
+            {
+                igraph_vs_destroy(&v_sel);
+                return ERR(igraph_strerror(res));
+            }
+
+            if (auto res = igraph_neighborhood(graph->get_graph(), &neighborhoods_raw, v_sel, order, mode, min_dist); res != IGRAPH_SUCCESS)
+            {
+                igraph_vs_destroy(&v_sel);
+                igraph_vector_int_list_destroy(&neighborhoods_raw);
+                return ERR(igraph_strerror(res));
+            }
+
+            std::vector<std::vector<u32>> neighborhoods;
+            for (u32 i = 0; i < igraph_vector_int_list_size(&neighborhoods_raw); i++)
```

Comment:
Another tip:

_Generally_, repeated calls to igraph's `_size` functions in a for loop can be detrimental to performance. This is because these C functions cannot be inlined, so the compiler generates a function call for each. The workaround is to call `igraph_vector_int_list_size()` _once_, save the result in a variable, and use that.

We do try to mitigate this situation, and annotate such functions with GCC's `__attribute__((pure))`, but this is an imperfect solution that only works with certain compilers.

Of course the performance impact won't be relevant in all scenarios, and is very likely irrelevant in this specific case.